# 🔢 NumPy – The Core of Scientific Computing in Python

- 🐍 **NumPy** is a Python library used for **scientific computing**.  
- 📦 It provides:
  - **Multidimensional arrays** (ndarray object)  
  - Extra objects like **masked arrays** and **matrices**  
  - Fast routines for:
    - ➕ Mathematical operations  
    - 🔍 Logical operations  
    - 🔄 Shape manipulation  
    - 📊 Sorting & selecting  
    - 💾 Input/Output (I/O)  
    - 🎵 Fourier transforms  
    - 📐 Linear algebra  
    - 📈 Statistics  
    - 🎲 Random simulations  

<br>

## 🌟 Core Concept
- At the heart of NumPy is the **ndarray object**.  
- 🧩 It represents **n‑dimensional arrays** (1D, 2D, 3D, …).  
- ⚡ All elements inside are of the **same data type** (homogeneous).  

<br>

### 📝 Easy Analogy
Think of **NumPy arrays** as **super‑powered lists**:
- Normal Python list → `[1, 2, 3]`  
- NumPy array → faster, more memory‑efficient, and supports **matrix/tensor operations** directly.

---

# 1. 🔄 `NumPy Arrays` vs `Python Sequences`

- 📏 **Fixed Size** → NumPy arrays have a fixed size when created.  
  - Python lists can grow/shrink dynamically.  
  - Changing size of a NumPy array creates a **new array** and deletes the old one.  

- 🧩 **Same Data Type** → All elements in a NumPy array must be of the **same type**.  
  - This makes them equal in memory size and faster to process.  

- ⚡ **Efficient Operations** → NumPy arrays allow advanced math & data operations.  
  - These run **faster** and with **less code** compared to Python lists.  

- 📚 **Widely Used** → Many scientific & mathematical Python packages use NumPy arrays.  
  - Even if they accept Python lists, they **convert them into NumPy arrays** internally.  
  - Most of them also output results as NumPy arrays.

  <br>
  <br>

### 📝 Exam‑style Takeaway
- **NumPy arrays** = fixed size, homogeneous type, optimized for math/science.  
- **Python sequences (lists/tuples)** = flexible size, mixed types, slower for large computations.  


---


## 2. 📘 Structural Relationship of NumPy Arrays

| **Array Dimension** | **Name**   | **Rank** |
|----------------------|------------|----------|
| 1D Array             | Vector     | Rank 1   |
| 2D Array             | Matrix     | Rank 2   |
| 3D Array             | Tensor     | Rank 3   |


<br>

### 📝 Exam‑style Notes
- **1D (Vector)** → A simple list of numbers, e.g. `[1, 2, 3]`.  
- **2D (Matrix)** → A grid of rows and columns, e.g. `[[1, 2], [3, 4]]`.  
- **3D (Tensor)** → A cube‑like structure with depth, e.g. `[[[1,2],[3,4]], [[5,6],[7,8]]]`.  

<br>

---


### 1. Creating Numpy Arrays

In [ ]:
import numpy as np

In [ ]:
# np.array
import numpy as np

x  = [1,2,3]
print(type(x))

print()

a = np.array([1,2,3])
print(type(a))
print(a)

<class 'list'>

<class 'numpy.ndarray'>
[1 2 3]


In [ ]:
# 2D and 3D
b = np.array([[1,2,3],[4,5,6]])
print(b)

[[1 2 3]
 [4 5 6]]


In [ ]:
# 3D - Tensor
c = np.array([[[1,2],[3,4]],[[5,6],[7,8]]])
print(c)

[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


---

## 2. 🔍 What is `Dtype`

`dtype`=float tells NumPy to store the array elements as floating‑point numbers instead of integers.

### 📝  Notes on `dtype`
- **`dtype`** = data type of elements in a NumPy array.
- Common options:
  - `int` → integers (`int32`, `int64` depending on system).
  - `float` → floating‑point numbers (`float32`, `float64`).
  - `complex` → complex numbers.
  - `bool` → Boolean values (`True/False`).
  - `str` → strings.
- You can check with `arr.dtype` and change with `astype()`.

In [ ]:
# dtype: Float
np.array([1,2,3],dtype=float)

array([1., 2., 3.])

In [ ]:
# dtype: Int32
np.array([1,2,3],dtype=np.int32)

array([1, 2, 3], dtype=int32)

---

## 🔍 3. What is `np.arange`
- It’s a **NumPy function** used to create arrays with evenly spaced values.  
- Similar to Python’s built‑in `range()`, but returns a **NumPy array** instead of a list.  

### 📝 Syntax
```python
numpy.arange([start, ]stop, [step], dtype=None)
```

- **start** → (optional) starting value (default = 0).  
- **stop** → end value (exclusive).  
- **step** → difference between values (default = 1).  
- **dtype** → data type of array elements (optional).  


In [ ]:

# Example 1: Only stop
arr1 = np.arange(5)
print(arr1)   # [0 1 2 3 4]

# Example 2: Start and stop
arr2 = np.arange(2, 10)
print(arr2)   # [2 3 4 5 6 7 8 9]

# Example 3: Start, stop, step
arr3 = np.arange(1, 10, 2)
print(arr3)   # [1 3 5 7 9]

# Example 4: With dtype
arr4 = np.arange(1, 5, dtype=float)
print(arr4)   # [1. 2. 3. 4.]


[0 1 2 3 4]
[2 3 4 5 6 7 8 9]
[1 3 5 7 9]
[1. 2. 3. 4.]


---

## 4. 🔍 What is `reshape`
- `reshape` changes the **shape (dimensions)** of a NumPy array **without changing its data**.  
- You can reorganize the same elements into rows, columns, or higher dimensions.  

<br>

### 📝 Syntax
```python
numpy.reshape(a, newshape)
```
- **a** → the original array.  
- **newshape** → tuple defining the new dimensions (must match total number of elements).  

In [ ]:

# Example 1: 1D → 2D
print("###Example 1: 1D → 2D")
arr = np.arange(6)          # [0 1 2 3 4 5]
reshaped = arr.reshape(2, 3)
print(reshaped)

print()
print()

# Example 2: 1D → 3D
print("###Example 2: 1D → 3D")
arr = np.arange(8)          # [0 1 2 3 4 5 6 7]
reshaped = arr.reshape(2, 2, 2)
print(reshaped)


print()
print()


# Example 3: Using -1 (auto dimension)
print("###Example 3: Using -1 (auto dimension)")
arr = np.arange(12)
reshaped = arr.reshape(3, -1)   # NumPy auto-calculates columns
print(reshaped)



###Example 1: 1D → 2D
[[0 1 2]
 [3 4 5]]


###Example 2: 1D → 3D
[[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]]


###Example 3: Using -1 (auto dimension)
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


---

## 5. 🪸 `np.ones(shape, dtype=None)`

- `shape` → tuple defining dimensions (e.g., (3,4) = 3 rows × 4 columns).
- `dtype` → optional, sets type of elements (default = float).

In [ ]:
# np.ones and np.zeros
np.ones((3,4),dtype=np.int32)

array([[1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1]], dtype=int32)

In [ ]:
np.ones((3,4))

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

---


## 6. 🔍  `np.zeros((3,4))`
- `np.zeros((3,4))` creates a **3×4 array** filled with `0.0` (default dtype = `float64`).  
- Output:
  ```
  [[0. 0. 0. 0.]
   [0. 0. 0. 0.]
   [0. 0. 0. 0.]]
  ```
- `arr.dtype` → `float64` (unless you specify another dtype).


### 📝 Exam‑style Notes
- **Function**: `np.zeros(shape, dtype=None)`  
  - `shape` → tuple defining dimensions (e.g., `(3,4)` = 3 rows × 4 columns).  
  - `dtype` → optional, sets type of elements (default = float).  



In [ ]:
np.zeros((3,4))

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

---



## 7.  🔍  `np.random.random((3,4))`
- `np.random.random((3,4))` creates a **3×4 array** filled with random numbers.  
- Each number is a **float between 0.0 and 1.0** (uniform distribution).  
- Example output (your values will differ each run):
  ```
  [[0.374 0.892 0.123 0.456]
   [0.789 0.222 0.678 0.999]
   [0.111 0.555 0.333 0.777]]
  ```



### 📝 Exam‑style Notes
- **Function**: `np.random.random(shape)`  
  - `shape` → tuple defining dimensions (e.g., `(3,4)` = 3 rows × 4 columns).  
- **Range**: values ∈ [0.0, 1.0).  
- **dtype**: always float.  


### 📘 Related Random Functions
- `np.random.randint(low, high, size)` → random integers.  
- `np.random.rand(m, n)` → random floats in [0,1), similar to `random((m,n))`.  
- `np.random.randn(m, n)` → random samples from a **normal distribution**.  
- `np.random.uniform(low, high, size)` → random floats in a custom range.  



In [ ]:
# np.random
np.random.random((3,4))

array([[0.28080885, 0.27359044, 0.27813478, 0.82128296],
       [0.52689853, 0.9708459 , 0.07675664, 0.59877218],
       [0.5427885 , 0.8093251 , 0.82903356, 0.99003666]])

---


## 🔍 8.  `np.linspace`
- It generates an array of **evenly spaced numbers** over a specified interval.  
- Unlike `np.arange` (which uses step size), `np.linspace` uses the **number of samples** you want.  



## 📝 Syntax
```python
numpy.linspace(start, stop, num=50, dtype=None)
```
- **start** → starting value.  
- **stop** → ending value (inclusive).  
- **num** → number of values to generate (default = 50).  
- **dtype** → optional, sets data type.  





In [ ]:
# np.linspace
np.linspace(-10,10,10,dtype=int)

array([-10,  -8,  -6,  -4,  -2,   1,   3,   5,   7,  10])

---

## 9. 🔍  `np.identity(3)`
- `np.identity(3)` creates a **3×3** identity matrix.
- An identity matrix has 1s on the main diagonal and 0s everywhere else.
- Default `dtype` = `float64` (you can change it with dtype=int if needed).

In [ ]:
# np.identity
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

---

##  10. Array Attributes


### 🔍 What happens `np.arange(10)`
- `np.arange(10)` → generates numbers from **0 to 9** (stop is exclusive).  
- `dtype=np.int32` → stores each element as a **32‑bit integer**.  
- Output:
  ```
  [0 1 2 3 4 5 6 7 8 9]
  ```
- `a1.dtype` → `int32`.



#### 📝 Exam‑style Notes
- **`np.arange(start, stop, step, dtype)`**  
  - `start` → default = 0.  
  - `stop` → end value (exclusive).  
  - `step` → default = 1.  
  - `dtype` → controls element type (here `int32`).  





In [ ]:
a1 = np.arange(10,dtype=np.int32)
a2 = np.arange(12,dtype=float).reshape(3,4)
a3 = np.arange(8).reshape(2,2,2)



a3

array([[[0, 1],
        [2, 3]],

       [[4, 5],
        [6, 7]]])

### A. `ndim:`

- `a3.ndim` gives the number of dimensions (axes) of the array.

- In this example:
  - Shape = (2,3,4) → 3 dimensions.
  - So `a3.ndim` → 3.

In [ ]:
# ndim
a3.ndim

3

### B. `.shape`
- .shape returns a tuple.
- Each number in the tuple = size of that dimension.
- Examples:
  - `1D array` [1,2,3] → shape (3,).
  - `2D array` [[1,2],[3,4]] → shape (2,2).
  - `3D array` with shape (2,3,4) → 2 blocks, each of size 3×4.

In [ ]:
# shape
print(a3.shape)
a3

(2, 2, 2)


array([[[0, 1],
        [2, 3]],

       [[4, 5],
        [6, 7]]])


### C. `.size:`

- `a2.size` gives the **total number of elements** in the array.
- Here, a2 has shape `(3,4)` → 3 rows × 4 columns = **12 elements.**

In [ ]:
# size
print(a2.size)
a2

12


array([[ 0.,  1.,  2.,  3.],
       [ 4.,  5.,  6.,  7.],
       [ 8.,  9., 10., 11.]])

### D. `.itemsize` 
- `.itemsize` → memory size of one element in the array.
- Depends on the dtype:
  - int8 → 1 byte
  - int16 → 2 bytes
  - int32 → 4 bytes
  - int64 → 8 bytes
  - float32 → 4 bytes
  - float64 → 8 bytes

In [ ]:
# itemsize
a3.itemsize

8

### E. `a1.dtype`
- `a1.dtype` shows the data type of elements in the array.

In [ ]:
# dtype
print(a1.dtype)
print(a2.dtype)
print(a3.dtype)

int32
float64
int64


---

## ♻️ 11. Changing Datatype - `astype`:

- `.astype(np.int32)` → converts the `array’s elements` to the `specified data type` (int32 here).
- If `a3` was already integers, nothing changes except ensuring they are stored as 32‑bit integers.

In [ ]:
# astype
a3.astype(np.int32)

array([[[0, 1],
        [2, 3]],

       [[4, 5],
        [6, 7]]], dtype=int32)

---

## 12.  Array Operations

- `arange(n)` → sequence from `0` to `n-1`.
- `reshape(r,c)` → must satisfy `r × c = total elements`.
  - Ex: Here, 3 × 4 = 12, so reshape works perfectly.`

In [ ]:
a1 = np.arange(12).reshape(3,4)
a2 = np.arange(12,24).reshape(3,4)

print ("a1:",a1)
print()
print ("a2:",a2)

a1: [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]

a2: [[12 13 14 15]
 [16 17 18 19]
 [20 21 22 23]]


### A. 🔢 `Scalar Operations` in NumPy:
A `scalar operation` means applying a single number (scalar) or arithmetic operator to every element of an array.

<br>

| Operation | Example | Result |
| --- | --- | --- |
| Addition | `a1 `+ `2` | `[2, 3, 4, 5, 6, 7]` |
| Subtraction | `a1 `- `1` | `[-1, 0, 1, 2, 3, 4]`|
| Multiplication | `a1`* `3` | `[0, 3, 6, 9, 12, 15]` |
| Division | `a1`/ `2` | `[0. , 0.5, 1. , 1.5, 2. , 2.5]` |
| Power | `a1`** `2` | `[0, 1, 4, 9, 16, 25]` |

In [ ]:
# scalar operations

# arithmetic
a1 ** 2

array([[  0,   1,   4,   9],
       [ 16,  25,  36,  49],
       [ 64,  81, 100, 121]])

### B. 🧮 `Relational Operation` in NumPy
Relational operators (`==`, `!=`, `<`, `>`, `<=`, `>=`) compare each element of the array with a given value, producing a **Boolean array**.

Example:
```python
import numpy as np

a2 = np.arange(20, 30)   # [20, 21, 22, 23, 24, 25, 26, 27, 28, 29]
print(a2 == 15)
```

##### 📘 Output:
```
[False False False False False False False False False False]
```

##### ✨ Explanation:
- `a2 == 15` → checks if each element of `a2` is equal to 15.  
- Since `a2` contains values from 20 to 29, **none match 15**, so all results are `False`.  
- If `a2` had included 15, the corresponding position would be `True`.

| Operator | Example | Result (for `a2 = [10, 15, 20]`) |
|----------|---------|----------------------------------|
| Equal `==` | `a2 == 15` | `[False, True, False]` |
| Not Equal `!=` | `a2 != 15` | `[True, False, True]` |
| Greater `>` | `a2 > 15` | `[False, False, True]` |
| Less `<` | `a2 < 15` | `[True, False, False]` |



In [ ]:
# relational
a2 == 15

array([[False, False, False,  True],
       [False, False, False, False],
       [False, False, False, False]])

### C. 📝 `Vector Opertion`:
- Vector operations in NumPy are **element-wise** (pairwise across arrays of the same shape).
- If shapes differ, NumPy applies **broadcasting rules**.
- Works for `+`, `-`, `*`, `/`, `**`, etc.

### 🔍 Step-by-step
- `a1 = [0, 1, 2]`
- `a2 = [3, 4, 5]`
- Operation: `a1 ** a2` → element-wise power:
  - \(0^3 = 0\)  
  - \(1^4 = 1\)  
  - \(2^5 = 32\)  

So the result is `[0, 1, 32]`.

| Operation | Example | Result |
|-----------|---------|--------|
| Addition  | `a1 + a2` | `[3, 5, 7]` |
| Subtraction | `a1 - a2` | `[-3, -3, -3]` |
| Multiplication | `a1 * a2` | `[0, 4, 10]` |
| Division | `a1 / a2` | `[0. , 0.25, 0.4]` |
| Power | `a1 ** a2` | `[0, 1, 32]` |



In [ ]:
# vector operations
a1 ** a2

array([[                   0,                    1,                16384,
                    14348907],
       [          4294967296,         762939453125,      101559956668416,
           11398895185373143],
       [ 1152921504606846976, -1261475310744950487,  1864712049423024128,
         6839173302027254275]])

---

## 13. 🏘️ Array Functions

In [ ]:
a1 = np.random.random((3,3))
a1 = np.round(a1*100)
a1

array([[94., 32., 87.],
       [77., 77., 42.],
       [ 4., 97., 70.]])



### A. 📝 Array Function `(prod/sum/max/min)`:

- `axis=0` → operate **down columns.**
- `axis=1` → operate **across rows.**
- If no axis is given → operation applies to **all elements.**

#### ✅ Takeaway:
- `prod` → multiplication
- `sum` → addition
- `max` → `largest element`
- `min` → `smallest element`
- Axis decides direction (0 = column, 1 = row).

#### ✅ Example:
```
[[1 2 3]
 [4 5 6]]
```

<br>

| Function | Axis=0 (Column) | Axis=1 (Row) |
| --- | --- | --- |
| **prod** | ``[4, ``10, ``18]`` | ``[6, ``120]`` |
| **max** | ``[4, ``5, ``6]`` | ``[3, ``6]`` |
| **min** | ``[1, ``2, ``3]`` | ``[1, ``4]`` |
| **sum** | ``[5, ``7, ``9]`` | ``[6, ``15]`` |

In [ ]:
# max/min/sum/prod
# 0 -> col and 1 -> row

a1 = np.arange(1, 7).reshape(2, 3)
print("Matrix:\n", a1)

print()

print("Column-wise product:", np.prod(a1, axis=0))
print("Row-wise product:", np.prod(a1, axis=1))

print()

print("Column-wise max:", np.max(a1, axis=0))
print("Row-wise max:", np.max(a1, axis=1))

print()

print("Column-wise min:", np.min(a1, axis=0))
print("Row-wise min:", np.min(a1, axis=1))

print()

print("Column-wise sum:", np.sum(a1, axis=0))
print("Row-wise sum:", np.sum(a1, axis=1))


Matrix:
 [[1 2 3]
 [4 5 6]]

Column-wise product: [ 4 10 18]
Row-wise product: [  6 120]

Column-wise max: [4 5 6]
Row-wise max: [3 6]

Column-wise min: [1 2 3]
Row-wise min: [1 4]

Column-wise sum: [5 7 9]
Row-wise sum: [ 6 15]


### B. 📝 Array Function `(mean/median/std/var)`:

**Matrix `a1`:**
```
[[1 2 3]
 [4 5 6]]
```


#### 📘 Results

| Function | Axis=0 (Column) | Axis=1 (Row) |
|----------|-----------------|--------------|
| **mean**   | `[2.5, 3.5, 4.5]` | `[2.0, 5.0]` |
| **median** | `[2.5, 3.5, 4.5]` | `[2.0, 5.0]` |
| **std**    | `[1.5, 1.5, 1.5]` | `[0.816, 0.816]` |
| **var**    | `[2.25, 2.25, 2.25]` | `[0.667, 0.667]` |



#### 📝 Exam-style Notes
- **mean** → average of elements.  
- **median** → middle value.  
- **std** → standard deviation (spread).  
- **var** → variance (square of std).  
- `axis=0` → column-wise, `axis=1` → row-wise.  
- If no axis → operation applies to **all elements**.



In [ ]:
# mean/median/std/var

a1 = np.arange(1, 7).reshape(2, 3)
print("Matrix:\n", a1)

print()

print("Column-wise mean:", np.mean(a1, axis=0))
print("Row-wise mean:", np.mean(a1, axis=1))

print()

print("Column-wise median:", np.median(a1, axis=0))
print("Row-wise median:", np.median(a1, axis=1))

print()

print("Column-wise std:", np.std(a1, axis=0))
print("Row-wise std:", np.std(a1, axis=1))

print()

print("Column-wise var:", np.var(a1, axis=0))
print("Row-wise var:", np.var(a1, axis=1))


Matrix:
 [[1 2 3]
 [4 5 6]]

Column-wise mean: [2.5 3.5 4.5]
Row-wise mean: [2. 5.]

Column-wise median: [2.5 3.5 4.5]
Row-wise median: [2. 5.]

Column-wise std: [1.5 1.5 1.5]
Row-wise std: [0.81649658 0.81649658]

Column-wise var: [2.25 2.25 2.25]
Row-wise var: [0.66666667 0.66666667]


### C. `np.sin(x)`
- `np.sin(x)` → computes the `sine` of each element (in radians).

<br>

| Function | Example | Result |
| --- | --- | --- |
| ``np.sin(a1)`` | ``[0, ``0.841, ``0.909, ``0.141, ``-0.757]`` | Sine values |
| ``np.cos(a1)`` | ``[1, ``0.540, ``-0.416, ``-0.990, ``-0.654]`` | Cosine values |
| ``np.tan(a1)`` | ``[0, ``1.557, ``-2.185, ``-0.142, ``1.158]`` | Tangent values |

In [ ]:
# trigonomoetric functions
np.sin(a1)

array([[ 0.84147098,  0.90929743,  0.14112001],
       [-0.7568025 , -0.95892427, -0.2794155 ]])

### D. `np.dot(x, y)`
- `np.dot(x, y)` → performs dot product (matrix multiplication or inner product depending on dimensions).
- If both are `1D arrays` → it’s the `inner product` (sum of element‑wise multiplication).
- If they are `2D arrays` → it’s `matrix multiplication`.


In [ ]:
# dot product
a2 = np.arange(12).reshape(3,4)
a3 = np.arange(12,24).reshape(4,3)

np.dot(a2,a3)

array([[114, 120, 126],
       [378, 400, 422],
       [642, 680, 718]])

### E. `np.exp(x)`

- `np.exp(x)` → computes the **exponential function** \( e^x \) for each element.
- `np.exp()` is **vectorized** → applies to all elements without loops.
- Useful in:
  - **Machine Learning** → activation functions (softmax, sigmoid).
  - **Statistics** → exponential distributions.
  - **Finance** → compound interest calculations.

| Function | Example | Meaning |
| --- | --- | --- |
| ``np.exp(a1)`` | ``[e^0, ``e^1, ``e^2, ``...]`` | Exponential growth |
| ``np.log(a1)`` | Natural log | Inverse of exp |
| ``np.exp2(a1)`` | ``[2^0, ``2^1, ``2^2, ``...]`` | Power of 2 |

In [ ]:
# log and exponents
np.exp(a1)

array([[  2.71828183,   7.3890561 ,  20.08553692],
       [ 54.59815003, 148.4131591 , 403.42879349]])



### F. 🔍  `np.log(x)` 
- `np.log(x)` → computes the **natural logarithm** (base \(e\)) of each element.
- Applied element‑wise:
  - \(\log(0)\) → **undefined**, NumPy shows `-inf`.
  - \(\log(1) = 0\).
  - \(\log(2) \approx 0.693\).
  - \(\log(3) \approx 1.099\).
  - \(\log(4) \approx 1.386\).



### 📝 Exam-style Notes
- `np.log()` → natural log (base \(e\)).
- `np.log10()` → log base 10.
- `np.log2()` → log base 2.
- ⚠️ **Important**: `log(0)` is undefined → NumPy returns `-inf`.

| Function | Example | Result |
|----------|---------|--------|
| `np.log(a1)` | `[ -inf, 0, 0.693, 1.099, 1.386 ]` | Natural log |
| `np.log10(a1)` | `[ -inf, 0, 0.301, 0.477, 0.602 ]` | Base 10 log |
| `np.log2(a1)` | `[ -inf, 0, 1, 1.585, 2 ]` | Base 2 log |





### F. 🔍  `np.log(x)` 
- `np.log(x)` → computes the **natural logarithm** (base \(e\)) of each element.
- Applied element‑wise:
  - \(\log(0)\) → **undefined**, NumPy shows `-inf`.
  - \(\log(1) = 0\).
  - \(\log(2) \approx 0.693\).
  - \(\log(3) \approx 1.099\).
  - \(\log(4) \approx 1.386\).



### 📝 Exam-style Notes
- `np.log()` → natural log (base \(e\)).
- `np.log10()` → log base 10.
- `np.log2()` → log base 2.
- ⚠️ **Important**: `log(0)` is undefined → NumPy returns `-inf`.

| Function | Example | Result |
|----------|---------|--------|
| `np.log(a1)` | `[ -inf, 0, 0.693, 1.099, 1.386 ]` | Natural log |
| `np.log10(a1)` | `[ -inf, 0, 0.301, 0.477, 0.602 ]` | Base 10 log |
| `np.log2(a1)` | `[ -inf, 0, 1, 1.585, 2 ]` | Base 2 log |



In [ ]:
np.log(a1)

array([[0.        , 0.69314718, 1.09861229],
       [1.38629436, 1.60943791, 1.79175947]])

In [ ]:
# round/floor/ceil

np.ceil(np.random.random((2,3))*100)

array([[74., 77., 82.],
       [ 2., 68., 17.]])

---

## 💊 14. Indexing :

- `a[i,j]` → element at row `i`, column `j`.
- ✨✨✨ Indexing starts at **0**.
- `axis=0` → rows, `axis=1` → columns.
- 🍀 Negative indices work too: `a2[-1,0]` → last row, first column.

<br>

| Expression | Meaning | Result |
| --- | --- | --- |
| ``a2[0,0]`` | 1st row, 1st col | ``0`` |
| ``a2[0,2]`` | 1st row, 3rd col | ``2`` |
| ``a2[1,1]`` | 2nd row, 2nd col | ``4`` |
| ``a2[1,0]`` | 2nd row, 1st col | ``3`` |

In [ ]:
a1 = np.arange(10)
a2 = np.arange(12).reshape(3,4)
a3 = np.arange(8).reshape(2,2,2)

In [ ]:
print("a1:")
a1

a1:


array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
print("a2:")
a2

a2:


array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
print("a3:")
a3

a3:


array([[[0, 1],
        [2, 3]],

       [[4, 5],
        [6, 7]]])

##### 1. `a2[1,0]` → element at row `2`, column 1 → `3`.

In [ ]:
a2[1,0]

np.int64(4)

##### 2. 📘 Step-by-step Indexing `3D`


```
[
 [[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]
]

```
- `a3[1,0,1]` → 3D indexing:
- `1` → 2nd block (outermost dimension).
- `0` → 1st row inside that block.
- `1` → 2nd column inside that row.

So we go to:

- Block 2 → `[[12 13 14 15], [16 17 18 19], [20 21 22 23]]`
- Row 1 → `[12 13 14 15]`
- Column 2 → `13`

Result = 13

In [ ]:
a3[1,0,1]

np.int64(5)

In [ ]:
a3[1,1,0]

np.int64(6)

---

## 🏗️ 15. Slicing:


#### ✨ Explanation of Slice `[start:stop:step]`
- `start = 2` → begin at index 2 (value = 2).  
- `stop = 5` → go up to index 5 (exclusive).  
- `step = 2` → take every 2nd element.  

So from `[2, 3, 4]` → we pick `2` and `4`.

<br>

#### 📝 Exam-style Notes
- General form: `a[start:stop:step]`.  
- Default values:
  - `start` → 0 if omitted.  
  - `stop` → end of array if omitted.  
  - `step` → 1 if omitted.  
- Negative indices work too: `a1[-1] = 9`.  
- Negative step reverses: `a1[::-1] = [9 8 7 6 5 4 3 2 1 0]`.

<br>

| Expression | Meaning | Result |
|------------|---------|--------|
| `a1[2:5]` | from index 2 to 4 | `[2 3 4]` |
| `a1[2:5:2]` | from 2 to 4, step 2 | `[2 4]` |
| `a1[:4]` | first 4 elements | `[0 1 2 3]` |
| `a1[::2]` | every 2nd element | `[0 2 4 6 8]` |
| `a1[::-1]` | reversed array | `[9 8 7 6 5 4 3 2 1 0]` |



In [ ]:
print("a1:",a1)

print()

print(a1[2:5:2])

a1: [0 1 2 3 4 5 6 7 8 9]

[2 4]





**Matrix `a2`:**
```
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
```



### 📘 Step-by-step
- `0:2` → take **rows 0 and 1** (first two rows).  
- `1::2` → start at column index 1, then take every 2nd column.  
  - So columns = 1 and 3.  

Selected submatrix:
```
[[1 3]
 [5 7]]
```



### 📝 Exam-style Notes
- General form: `a[start_row:end_row:row_step, start_col:end_col:col_step]`.  
- First part before comma → row slice.  
- Second part after comma → column slice.  
- `start_row` → index of first row to include
- `end_row` → index where slicing stops (exclusive)
- `row_step` → step size for rows (`default = 1`)
- `start_col` → index of first column to include
- `end_col` → index where slicing stops (exclusive)
- `col_step` → step size for columns (`default = 1`)

| Expression | Meaning | Result |
|------------|---------|--------|
| `a2[0:2, 1::2]` | Rows 0–1, Cols 1 & 3 | `[[1,3],[5,7]]` |
| `a2[:, ::2]` | All rows, every 2nd col | `[[0,2],[4,6],[8,10]]` |
| `a2[1:, 2:]` | Rows 1–end, Cols 2–end | `[[6,7],[10,11]]` |



In [ ]:
print("a2:",a2)

print()

print(a2[0:2,1::2])

a2: [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]

[[1 3]
 [5 7]]




### 📘 General Formula - For a `2D array (matrix)`:


$a[start\_row:end\_row:row\_step,\; start\_col:end\_col:col\_step]$


- **start_row** → index of first row to include  
- **end_row** → index where slicing stops (exclusive)  
- **row_step** → step size for rows (default = 1)  
- **start_col** → index of first column to include  
- **end_col** → index where slicing stops (exclusive)  
- **col_step** → step size for columns (default = 1)  

| Slice | Meaning | Result |
|-------|---------|--------|
| `a2[::2, :]` | Every 2nd row (rows 0 & 2), all columns | `[[0 1 2 3],[8 9 10 11]]` |
| `a2[:, 1::2]` | All rows, start at col 1, step 2 (cols 1 & 3) | `[[1 3],[5 7],[9 11]]` |
| `a2[::2, 1::2]` | Every 2nd row (rows 0 & 2), cols 1 & 3 | `[[1 3],[9 11]]` |




In [ ]:
a2[::2,1::2]

array([[ 1,  3],
       [ 9, 11]])

### 📘 Slicing `Step-by-step:`
- `1 `→ select **row index 1** (the second row).
- `::3` → take every 3rd element from that row (step = 3).
Row 1 = `[4 5 6 7]`
- Index 0 → `4`
- Index 3 → `7`

In [ ]:
a2[1,::3]

array([4, 7])

### `a2[0,:]` → extracts the **first row** → `[0, 1, 2, 3]`.

In [ ]:
a2[0,:]

array([0, 1, 2, 3])

 ### `a2[:,2]` → extracts the **entire 3rd column** → `[2, 6, 10]`.

In [ ]:
a2[:,2]

array([ 2,  6, 10])

 ### `a2[1:, 1:3]`  → extracts the `Rows 1–end`, `Cols 1–2` → [[5,6],[9,10]]

In [ ]:
a2[1:,1:3]

array([[ 5,  6],
       [ 9, 10]])

### `Review a3`: TENSOR - `3D Array`
- `np.arange(27)` → numbers from **0 to 26** (27 elements).
- `.reshape(3,3,3)` → reorganizes into a **3 × 3 × 3 cube:**
  - 3 blocks (outer dimension).
  - Each block has 3 rows × 3 columns.
- So it’s a **tensor of shape (3,3,3).** 

In [ ]:
a3 = np.arange(27).reshape(3,3,3)
a3

array([[[ 0,  1,  2],
        [ 3,  4,  5],
        [ 6,  7,  8]],

       [[ 9, 10, 11],
        [12, 13, 14],
        [15, 16, 17]],

       [[18, 19, 20],
        [21, 22, 23],
        [24, 25, 26]]])

### 📘 General Formula - For a `3D array`:


$a[start\_block:end\_block:block\_step,\; start\_row:end\_row:row\_step,\; start\_col:end\_col:col\_step]$


- **start_block:end_block:block_step** → slice along the **depth (blocks)**.  
- **start_row:end_row:row_step** → slice along the **rows** inside each block.  
- **start_col:end_col:col_step** → slice along the **columns** inside each row.  





### 📝 Exam-style Notes
- General form: `a[block_slice, row_slice, col_slice]`.  
- Each dimension follows the same `[start:end:step]` rule.  
- Negative indices and steps work too.  
  - Example: `a3[::-1, :, :]` → reverse blocks.  
  - Example: `a3[:, ::-1, :]` → reverse rows.  
  - Example: `a3[:, :, ::-1]` → reverse columns.

| Slice | Meaning | Result |
|-------|---------|--------|
| `a3[0,:,:]` | Block 0, all rows & cols | `[[0 1 2],[3 4 5],[6 7 8]]` |
| `a3[:,1,:]` | All blocks, row 1 | `[[3 4 5],[12 13 14],[21 22 23]]` |
| `a3[:,:,2]` | All blocks, col 2 | `[[2 5 8],[11 14 17],[20 23 26]]` |
| `a3[0:2,1:,1:3]` | Blocks 0–1, rows 1–2, cols 1–2 | `[[[4 5],[7 8]],[[13 14],[16 17]]]` |


### `a3[::2,0,::2]`


| Slice | Meaning | Result |
| --- | --- | --- |
| ``a3[::2, ``:, ``:]`` | Every 2nd block (blocks 0 & 2), all rows & cols | ``[[[0 ``1 ``2],[3 ``4 ``5],[6 ``7 ``8]], ``[[18 ``19 ``20],[21 ``22 ``23],[24 ``25 ``26]]]`` |
| ``a3[::2, ``0, ``:]`` | From those blocks, take **row 0 only** | ``[[0 ``1 ``2],[18 ``19 ``20]]`` |
| ``a3[::2, ``0, ``::2]`` | From that row, take every 2nd column (cols 0 & 2) | ``[[0 ``2],[18 ``20]]`` |

In [ ]:
a3[::2,0,::2]

array([[ 0,  2],
       [18, 20]])

In [ ]:
a3[2,1:,1:]

array([[22, 23],
       [25, 26]])

In [ ]:
a3[0,1,:]

array([3, 4, 5])

---

## 16. 🐦‍🔥 Iterating

- `Loops` through rows of the matrix.

#### A. Loop through `1-D matrix`

In [ ]:
a1

for i in a1:
  print(i)

0
1
2
3
4
5
6
7
8
9


#### B. Loop through `2-D matrix`

##### 🔍 Step-by-step Table

| Step | Meaning | Result |
|------|---------|--------|
| `for i in a2:` | Loop through each row of `a2` | Iterates 3 times (row 0, row 1, row 2) |
| First iteration | `i = [0 1 2 3]` | Prints `[0 1 2 3]` |
| Second iteration | `i = [4 5 6 7]` | Prints `[4 5 6 7]` |
| Third iteration | `i = [8 9 10 11]` | Prints `[8 9 10 11]` |



##### 📝 Key Notes
- Iterating a **2D array** → gives **rows** (1D arrays).  
- Iterating a **3D array** → gives **blocks** (2D arrays).  
- To iterate element by element:
  - `for i in a2.flat:` → prints each number one by one.  
  - `np.nditer(a2)` → advanced element-wise iteration.

In [ ]:
a2

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
for i in a2:
  print(i)

[0 1 2 3]
[4 5 6 7]
[ 8  9 10 11]


#### B. Loop through `3-D matrix`

##### 🔍 Step-by-step Table

| Step | Meaning | Result |
|------|---------|--------|
| `for i in a3:` | Loop through each **block** (outer dimension) | Iterates 3 times (block 0, block 1, block 2) |
| First iteration | `i = [[0 1 2],[3 4 5],[6 7 8]]` | Prints block 0 |
| Second iteration | `i = [[9 10 11],[12 13 14],[15 16 17]]` | Prints block 1 |
| Third iteration | `i = [[18 19 20],[21 22 23],[24 25 26]]` | Prints block 2 |



##### 📝 Key Notes
- Iterating a **2D array** → gives rows (1D arrays).  
- Iterating a **3D array** → gives blocks (2D arrays).  
- To iterate element by element:
  - `for i in a3.flat:` → prints each number one by one.  
  - `np.nditer(a3)` → advanced element-wise iteration.


##### ✅ **Takeaway**:  
`for i in a3:` → loops through **2D blocks** of the 3D tensor.  

In [ ]:
a3

array([[[ 0,  1,  2],
        [ 3,  4,  5],
        [ 6,  7,  8]],

       [[ 9, 10, 11],
        [12, 13, 14],
        [15, 16, 17]],

       [[18, 19, 20],
        [21, 22, 23],
        [24, 25, 26]]])

In [ ]:
for i in a3:
  print(i)

[[0 1 2]
 [3 4 5]
 [6 7 8]]
[[ 9 10 11]
 [12 13 14]
 [15 16 17]]
[[18 19 20]
 [21 22 23]
 [24 25 26]]


##### D.📘 What’s Happening `np.nditer`
- `np.nditer(a3)` → creates an iterator that goes through every `single element of the 3D` array `one by one`.
- Unlike for `i in a3`: (which gave you 2D blocks), this goes element-wise.

In [ ]:
for i in np.nditer(a3):
  print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26


---

## 17. ˚⊱🪷⊰˚Reshaping

In [ ]:
# reshape
np.arange(12).reshape(3,4)

In [ ]:
# Transpose
np.transpose(a2)
a2.T

array([[ 0,  4,  8],
       [ 1,  5,  9],
       [ 2,  6, 10],
       [ 3,  7, 11]])


###  17.1 What `ravel()` Does
- **Flattens** a multi‑dimensional array into **1D**.  
- Returns a **view** of the array (not a copy) whenever possible → meaning changes in the flattened array may reflect back in the original.  
- Faster and memory‑efficient compared to `flatten()` (which always returns a copy).  

In [ ]:
# ravel
a3.ravel()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26])

---

## 🧬 18. Stacking

In [ ]:
# Create two Matrix:
a4 = np.arange(12).reshape(3,4)
a5 = np.arange(12,24).reshape(3,4)
a5

array([[12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

###  🔍 `HSTACK` vs `VSTACK`
- **`np.hstack`** → stacks arrays **horizontally (column-wise)**.  
- **`np.vstack`** → stacks arrays **vertically (row-wise)**. 
- It requires that the arrays have the **same number of rows**.  
- The result is a new array where the columns of `a5` are appended to the columns of `a4`.



#### 📊 Comparison Table

| Function | Meaning | Example Result |
|----------|---------|----------------|
| `np.hstack((a4,a5))` | Horizontal stack (columns side by side) | `[[1 2 5 6],[3 4 7 8]]` |
| `np.vstack((a4,a5))` | Vertical stack (rows one below another) | `[[1 2],[3 4],[5 6],[7 8]]` |
| `np.concatenate((a4,a5), axis=1)` | Same as `hstack` | `[[1 2 5 6],[3 4 7 8]]` |
| `np.concatenate((a4,a5), axis=0)` | Same as `vstack` | `[[1 2],[3 4],[5 6],[7 8]]` |


#### ✅ **Takeaway**:  
- `np.hstack((a4,a5))` → joins arrays **side by side** (column-wise).  
- - `np.vstack((a4,a5))` → joins arrays **top to bottom** (row-wise).


In [ ]:
# horizontal stacking
np.hstack((a4,a5))

array([[ 0,  1,  2,  3, 12, 13, 14, 15],
       [ 4,  5,  6,  7, 16, 17, 18, 19],
       [ 8,  9, 10, 11, 20, 21, 22, 23]])

In [ ]:
# Vertical stacking
np.vstack((a4,a5))

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

---

## 🧬 19. Splitting

In [ ]:
# horizontal splitting
a4

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

### ✅ **Takeaway**:  
- `np.hsplit(a4, 2)` → divides the array into **2 equal column groups**.  
- `np.vsplit(a4, 2)` → divides the array into **2 equal row groups**.


In [ ]:
np.hsplit(a4,2)

[array([[0, 1],
        [4, 5],
        [8, 9]]),
 array([[ 2,  3],
        [ 6,  7],
        [10, 11]])]

### vertical splitting

In [ ]:
a5

array([[12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

In [ ]:
np.vsplit(a5,2)

ValueError: array split does not result in an equal division

---